.github/workflows/build.yml

```diff
@@ -0,0 +1,104 @@
+name: build
+
+on:
+  push:
+    branches:
+      - main
+      - release/**
+  workflow_dispatch:
+
+env:
+  PKG_NAME: 'terraform-cloud-operator'
+  DOCKER_HUB_REPO: 'hashicorp/terraform-cloud-operator'
+  DOCKER_HUB_DEV_REPO: 'hashicorppreview/terraform-cloud-operator'
+
+jobs:
+  set-product-version:
+    runs-on: ubuntu-latest
+    outputs:
+      product-version: ${{ steps.set-product-version.outputs.product-version }}
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@8f4b7f84864484a7bf31766abe9204da3cbe65b3 # v3.5.0
+      - name: Set product version
+        id: set-product-version
+        uses: hashicorp/actions-set-product-version@06bddb8ee70f0dda2e21c39b5f0913d36cc9d657 # v1
+
+  generate-metadata-file:
+    needs: set-product-version
+    runs-on: ubuntu-latest
+    env:
+      METADATA_FILE_NAME: 'metadata.json'
+    outputs:
+      filepath: ${{ steps.generate-metadata-file.outputs.filepath }}
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@8f4b7f84864484a7bf31766abe9204da3cbe65b3 # v3.5.0
+      - name: Generate metadata file
+        id: generate-metadata-file
+        uses: hashicorp/actions-generate-metadata@fdbc8803a0e53bcbb912ddeee3808329033d6357 # v1.1.1
+        with:
+          version: ${{ needs.set-product-version.outputs.product-version }}
+          product: ${{ env.PKG_NAME }}
+          repositoryOwner: 'hashicorp'
+          metadataFileName: ${{ env.METADATA_FILE_NAME }}
+      - uses: actions/upload-artifact@0b7f8abb1508181956e8e162db84b466c27e18ce # v3.1.2
+        with:
+          name: metadata.json
+          path: ${{ env.METADATA_FILE_NAME }}
+
+  build:
+    name: Go Linux ${{ matrix.arch }} build
+    needs: [set-product-version]
+    runs-on: ubuntu-latest
+    outputs:
+      go-version: ${{ steps.set-up-go.outputs.go-version }}
```

Comment:
This _might_ cause some unexpected behavior in the future. I'm still reading through the changes to determine if this is the case or not, but for now I'm just leaving this here as an informational-only comment.   

Previously, setting an output in a job followed a 'last one wins' rules, leading to the value of the output being non-deterministic. In this case that's not an issue since it should be a stable value.  However, they're adding functionality to be able to set the output name based on the properties of the matrix.  From what I've read so far, I don't think this will cause any problems with how it's being used here, but it's something I think it's worth being aware of. 

Reference:
https://github.com/orgs/community/discussions/17245
https://github.com/actions/runner/pull/2477

Reply:
Hi @smacfarlane,

Thank you for the valuable comment and references. This is indeed an interesting point.

As you can see, I don't have the `.go-version` file and fully rely on the `go.mod` and the `go-version-file` option of the `actions/setup-go` action. Basically, I am fine to have the latest patch of Go during the release. However, with what you have shared, there is still a tiny possibility that one runner get a newer version of Go.

Initially, I had one extra step `go-get-version` that only did one thing -- install Go with the `actions/setup-go` action, and then I reuse the output everywhere. Then I removed this step as 'unnecessary'. Now you make me think about this again.

This is all done because I don't want to have the `.go-version` file and want to rely only on `go.mod`. I am just afraid that at some point `.go-version` won't by sync with `go.mod`.

What do you think about this?

Thanks!